JustSaying/Messaging/MessageProcessingStrategies/Throttled.cs

```diff
@@ -3,81 +3,203 @@
 
-        public async Task StartWorker(Func<Task> action, CancellationToken cancellationToken)
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Throttled"/> class.
+        /// </summary>
+        /// <param name="options">The <see cref="ThrottledOptions"/> to use.</param>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="options"/> is <see langword="null"/>.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// <paramref name="options"/> does not specify an <see cref="ILogger"/> or <see cref="IMessageMonitor"/>.
+        /// </exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// The concurrency specified by <paramref name="options"/> is less than one,
+        /// or the start timeout specified by <paramref name="options"/> is zero or negative.
+        /// </exception>
+        public Throttled(ThrottledOptions options)
         {
-            try
+            if (options == null)
             {
-                await _semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+                throw new ArgumentNullException(nameof(options));
             }
-            catch (OperationCanceledException)
+
+            if (options.MaxConcurrency < 1)
             {
-                return;
+                throw new ArgumentOutOfRangeException(nameof(options), options.MaxConcurrency, "The maximum concurrency value must be a positive integer.");
             }
 
-            _ = Task.Run(() => ReleaseOnCompleted(action));
+            if (options.StartTimeout <= TimeSpan.Zero && options.StartTimeout != Timeout.InfiniteTimeSpan)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(options),
+                    options.StartTimeout,
+                    "The start timeout must be a positive value or Timeout.InfiniteTimeSpan.");
+            }
+
+            _messageMonitor = options.MessageMonitor ?? throw new ArgumentException($"A value for {nameof(ThrottledOptions.MessageMonitor)} must be specified.", nameof(options));
+            _logger = options.Logger ?? throw new ArgumentException($"A value for {nameof(ThrottledOptions.Logger)} must be specified.", nameof(options));
+
+            MaxConcurrency = options.MaxConcurrency;
+            StartTimeout = options.StartTimeout;
+
+            _processSequentially = options.ProcessMessagesSequentially;
+            _semaphore = new SemaphoreSlim(MaxConcurrency, MaxConcurrency);
         }
 
-        private async Task ReleaseOnCompleted(Func<Task> action)
+        /// <summary>
+        /// Finalizes an instance of the <see cref="Throttled"/> class.
+        /// </summary>
+        ~Throttled()
+        {
+            Dispose(false);
+        }
+
+        /// <summary>
+        /// Gets the available number of workers.
+        /// </summary>
+        public int AvailableWorkers => _semaphore.CurrentCount;
```

Comment:
Is this intentionally left in for monitoring purposes?

Reply:
I'm using it in the tests as well.